{% extends 'catalog/base.html' %}

{% block title %}Download Information{% endblock %}

{% block content %}
  <nav aria-label="You are here:">
    <ol class="breadcrumb">
      <li class="breadcrumb-item"><a href="../../">PGS Catalog</a></li>
      <li class="breadcrumb-item">Downloads</li>
    </ol>
  </nav>

    <div class="clearfix">
      <h2 class="float_left">PGS Catalog Downloads</h2>
      <div class="float_right">
        <a class="btn btn-pgs pgs_no_icon_link pgs-ftp-btn" href="{{ pgs_ftp_root }}" target="_blank">
          <span class="fa fa-folder"></span> PGS Catalog FTP site
        </a>
      </div>
    </div>
    <p>
        This page contains information about the PGS Catalog downloads and FTP site.
    </p>

    <section>
      <h3>Available PGS Catalog Downloads</h3>
      <table class="table table-bordered table_pgs_h">
        <tbody>
          <tr>
            <td>
              <span class="pgs_h_big_header">PGS Scoring Files</span>
              <a class="pgs_no_icon_link info-icon" href="#dl_ftp" data-toggle="tooltip" data-placement="bottom" title="Individual scoring and metadata files for each PGS. Click on this icon to see information about the PGS Scoring File format."><i class="fa fa-info-circle"></i></a>
            </td>
            <td><a class="btn btn-pgs pgs_no_icon_link pgs-ftp-btn" href="{{ pgs_ftp_root }}/scores/" target="_blank"><span class="fa fa-folder"></span> PGS Score FTP directories</a></td>
          </tr>
          <tr>
            <td>
              <span class="pgs_h_big_header">PGS Catalog Metadata</span>
              <a class="pgs_no_icon_link info-icon" href="#dl_ftp_metadata" data-toggle="tooltip" data-placement="bottom" title="Metadata files for all scores, publications, and traits in the PGS Catalog. Click on this icon to see information about the PGS Metadata File formats."><i class="fa fa-info-circle"></i></a>
            </td>
            <td><a class="btn btn-pgs pgs_no_icon_link pgs-ftp-btn" href="{{ pgs_ftp_root }}/metadata/" target="_blank"><span class="fa fa-folder"></span> PGS Metadata FTP directory</a></td>
          </tr>
        </tbody>
      </table>
    </section>

    <section>
      <h3 id="dl_ftp">PGS Catalog Scoring Files</h3>
        <p>Each score file is formatted to be a gzipped <b>tab-delimited</b> text file, labelled by its PGS Catalog Score ID (e.g. <code>PGS000001.txt.gz</code>).
        <p>In the current catalog the scores have been extracted from the relevant publication and only been reformatted to have a consistent header (lines starting with <code>#</code>)
           listing relevant PGS Catalog information, e.g.:
        </p>
        <div class="pgs_scoring_header">
          <pre >
<span>### PGS CATALOG SCORING FILE - see www.pgscatalog.org/downloads/#dl_ftp for additional information</span>
<span>## POLYGENIC SCORE (PGS) INFORMATION</span>
<b># PGS ID</b> = <i>PGS idenfier, e.g. 'PGS000001'</i>
<b># Reported Trait</b> = <i>trait, e.g. 'Breast Cancer'</i>
<b># Original Genome Build</b> = <i>Genome build/assembly, e.g. 'GRCh38'</i>
<b># Number of Variants</b> = <i>Number of variants listed in the PGS</i>
<span>## SOURCE INFORMATION</span>
<b># PGP ID</b> = <i>PGS publication idenfier, e.g. 'PGP000001'</i>
<b># Citation</b> = <i>Information about the publication</i>
<span class="scoring_col">rsID</span><span class="scoring_col">chr_name</span><span class="scoring_col">chr_position</span><span class="scoring_col">effect_allele</span><span class="scoring_col">reference_allele</span><span class="pgs_colour_2">...</span></pre>
      </div>

      <p>It also has been edited to have consistent column headings based on the following schema:</p>
      <table class="table table-striped table_pgs_auto">
        <thead class="thead-light">
          <tr><th>Column Header</th><th>Field Name</th><th>Field Description</th></tr>
        </thead>
        <tbody>
          <tr><td>rsID</td><td>dbSNP Accession ID (rsID) </td><td>The SNPâ€™s rs ID</td></tr>
          <tr><td>chr_name</td><td>Location - Chromosome </td><td>Chromosome name/number associated with the variant</td></tr>
          <tr><td>chr_position</td><td>Location - Position within the Chromosome</td><td>Chromosomal position associated with the variant</td></tr>
          <tr><td>effect_allele</td><td>Effect Allele</td><td>The allele that's dosage is counted (e.g. {0, 1, 2}) and multiplied by the variant's weight ('effect_weight') when calculating score. The effect allele is also known as the 'risk allele'.</td></tr>
          <tr><td>reference_allele</td><td>Reference Allele</td><td>The other allele(s) at the loci</td></tr>
          <tr><td>effect_weight</td><td>Variant Weight</td><td>Value of the effect that is multiplied by the dosage of the effect allele ('effect_allele') when calculating the score.</td></tr>
          <tr><td>locus_name</td><td>Locus Name</td><td>This is kept in for loci where the variant may be referenced by the gene (APOE e4). It is also common (usually in smaller PGS) to see the variants named according to the genes they impact.</td></tr>
          <tr><td>weight_type</td><td>Type of Weight</td><td>Whether the author supplied Variant Weight is a: <b>beta</b> (effect size), or something like an <b>OR/HR</b> (odds/hazard ratio)</td></tr>
          <tr><td>allelefrequency_effect</td><td>Effect Allele Frequency</td><td>Reported effect allele frequency, if the associated locus is a haplotype then haplotype frequency will be extracted.</td></tr>
          <tr><td>is_interaction</td><td>FLAG: Interaction</td><td>This is a TRUE/FALSE variable that flags whether the weight should be multiplied with the dosage of more than one variant. Interactions are demarcated with a <i>_x_</i> between entries for each of the variants present in the interaction. </td></tr>
          <tr><td>is_recessive</td><td>FLAG: Recessive Inheritance Model</td><td>This is a TRUE/FALSE variable that flags whether the weight should be added to the PGS sum only if there are 2 copies of the effect allele (e.g. it is a recessive allele).</td></tr>
          <tr><td>is_haplotype</i><br><i>is_diplotype</td><td>FLAG: Haplotype or Diplotype</td><td>This is a TRUE/FALSE variable that flags whether the effect allele is a haplotype/diplotype rather than a single SNP. Constituent SNPs in the haplotype are semi-colon separated. </td></tr>
          <tr><td>imputation_method</td><td>Imputation Method</td><td>This described whether the variant was specifically called with a specific imputation or variant calling method. This is mostly kept to describe HLA-genotyping methods (e.g. flag SNP2HLA, HLA*IMP) that gives alleles that are not referenced by genomic position.</td></tr>
          <tr><td>variant_description</td><td>Variant Description</td><td>This field describes any extra information about the variant (e.g. how it is genotyped or scored) that cannot be captured by the other fields.</td></tr>
          <tr><td>inclusion_criteria</td><td>Score Inclusion Criteria</td><td>Explanation of when this variant gets included into the PGS (e.g. if it depends on the results from other variants).</td></tr>
        </tbody>
      </table>
    </section>

    <section>
      <h3 id="dl_ftp_metadata">PGS Metadata File</h3>
      <p>Downloads of the full PGS Catalog metadata (data fields are described in more detail <a href="../docs/">here</a>), consists of an Excel file (<code>.xlsx</code>) containing all worksheets or an individual <code>.csv</code> file for each of the follwoing sheets:</p>
      <table class="table table-striped table_pgs_auto">
        <thead class="thead-light">
          <tr><th>Worksheet Name</th><th>Description</th></tr>
        </thead>
        <tbody>
          <tr><td>Readme</td><td>Information about the current PGS Catalog release.</td></tr>
          <tr><td>Scores</td><td>Lists and describes all PGS scores (excluding samples used for training, found in the <i>Score Development Samples</i> sheet).</td></tr>
          <tr><td>Performance Metrics</td><td>Lists the performance metrics and sources for each PGS evaluation in the catalog (information about the samples that were evaluated is found in the <i>Evaluation Sample Sets</i> sheet).</td></tr>
          <tr><td>Evaluation Sample Sets</td><td>Describes the samples used to evaluate PGS from the <i>Performance Metrics</i> sheet.</td></tr>
          <tr><td>Score Development Samples</td><td>Lists the source GWAS studies used to identify trait-associated variants, also describes the samples used in the GWAS studies and/or to train the PGS.</td></tr>
          <tr><td>Publications</td><td>Lists information about the Publications present in the GWAS Catalog.</td></tr>
          <tr><td>EFO Traits</td><td>Lists the Traits that are present in the PGS Catalog according to the EFO Ontology.</td></tr>
        </tbody>
      </table>
    </section>
{% endblock %}
